model/channel/channel.go

```diff
@@ -388,243 +361,274 @@ type SelectCase[T any] struct {
 	case3 *SelectCase[T3],
 	case4 *SelectCase[T4],
 	blocking bool) uint64 {
 
-	// Create an empty case with nil channel for the unused slot
-	emptyCase5 := &SelectCase[uint64]{}
+	i := primitive.RandomUint64() % uint64(4)
+	var selected bool
+	selected = TrySelectCase4(i, case1, case2, case3, case4)
+	if selected {
+		return i
+	}
 
-	return multiSelect(
-		case1,
-		case2,
-		case3,
-		case4,
-		emptyCase5,
-		blocking)
+	for {
+		selected = TrySelect(case1)
+		if selected {
+			return 0
+		}
+		selected = TrySelect(case2)
+		if selected {
+			return 1
+		}
+		selected = TrySelect(case3)
+		if selected {
+			return 2
+		}
+		selected = TrySelect(case4)
+		if selected {
+			return 3
+		}
+		if !blocking {
+			return 4
+		}
+	}
 }
 
-// Select5 is just an alias to MultiSelect for consistency in the API.
-func Select5[T1, T2, T3, T4, T5 any](
+func TrySelectCase5[T1, T2, T3, T4, T5 any](
+	index uint64,
 	case1 *SelectCase[T1],
 	case2 *SelectCase[T2],
 	case3 *SelectCase[T3],
 	case4 *SelectCase[T4],
-	case5 *SelectCase[T5],
-	blocking bool) uint64 {
-
-	return multiSelect(
-		case1,
-		case2,
-		case3,
-		case4,
-		case5,
-		blocking)
-}
-
-// Uses the applicable Try<Operation> function on the select case's channel. Default is always
-// selectable so simply returns true.
-func TrySelect[T any](select_case *SelectCase[T]) bool {
-	var channel *Channel[T] = select_case.channel
-	if channel == nil {
-		return false
+	case5 *SelectCase[T5]) bool {
+	if index == 0 {
+		return TrySelect(case1)
 	}
-	if select_case.dir == SelectSend {
-		return channel.TrySend(select_case.Value)
+	if index == 1 {
+		return TrySelect(case2)
 	}
-	if select_case.dir == SelectRecv {
-		var item T
-		var ok bool
-		var selected bool
-		selected, item, ok = channel.TryReceive()
-		// We can use these values for return by reference and they will be implicitly kept alive
-		// by the garbage collector so we can use value here for both the send and receive
-		// variants. What a miracle it is to not be using C++.
-		select_case.Value = item
-		select_case.Ok = ok
-		return selected
-
+	if index == 2 {
+		return TrySelect(case3)
 	}
-	return false
+	if index == 3 {
+		return TrySelect(case4)
+	}
+	if index == 4 {
+		return TrySelect(case5)
+	}
+	panic("index needs to be 0, 1, 2, 3 or 4")
 }
 
-func NewSendCase[T any](channel *Channel[T], value T) *SelectCase[T] {
-	return &SelectCase[T]{
-		channel: channel,
-		dir:     SelectSend,
-		Value:   value,
+func Select5[T1, T2, T3, T4, T5 any](
+	case1 *SelectCase[T1],
+	case2 *SelectCase[T2],
+	case3 *SelectCase[T3],
+	case4 *SelectCase[T4],
+	case5 *SelectCase[T5],
+	blocking bool) uint64 {
+
+	i := primitive.RandomUint64() % uint64(5)
+	var selected bool
+	selected = TrySelectCase5(i, case1, case2, case3, case4, case5)
+	if selected {
+		return i
 	}
-}
 
-func NewRecvCase[T any](channel *Channel[T]) *SelectCase[T] {
-	return &SelectCase[T]{
-		channel: channel,
-		dir:     SelectRecv,
+	for {
+		selected = TrySelect(case1)
```

Comment:
I think you might as well delete this variable and just write `if TrySelect(case1)`, though it won't make much difference in the proof due to `wp_auto`.